feat(tabs): Introduce Tabs component

andybraren opened the pull request:
This is an initial implementation of Tabs that resolves #30. The design spec for tabs is available [here](https://github.com/patternfly/patternfly-next/files/2357300/PF4_Tabs.pdf).

## Implementation notes

1. The scrollbar isn't hidden. We'll need to discuss this. (**EDIT:** We did in #859. We're keeping the scrollbar.)

2. The first example includes sample HTML for the `section`s that the tabs would control with the appropriate `role`, `id`, and `aria-labelledby` attributes. Is this fine? (**EDIT:** Yes.)

## CSS notes

1. The organization of the `SCSS` file could probably be improved. Tabs, secondary tabs, and the hover/focus/active states for each are all kinda interconnected.

2. Some modifiers like `pf-m-tabs-scrollLeftVisible` are highly specific to Tabs so I included "`tabs`" in their name. I’m guessing this should be changed to some existing convention? (**EDIT:** Per @mcoker's recommendation, they're now `.pf-m-start` and '.pf-m-end`)

3. Primary tabs need a 3px blue line above them when focused. Removing the existing `border-top` would make the tab’s text jump, so `border-top-color` is changed along with adding a 2px `box-shadow` to achieve the desired effect. It works, but maybe there's a cleaner way? (EDIT: Yes, using `::before` and `::after`, which is now implemented)

## Questions

1. A parent container is necessary whenever secondary tabs are used in order to position the fixed scroll buttons correctly (see the example). Should this parent container become a requirement, even when secondary tabs aren’t used? (**EDIT:** This component no longer needs a container. Primary and secondary tabs are now independent, and secondary receives extra styling only when directly adjacent to primary.)

2. The hover and focus states weren’t fully specified in the design spec. The contrast of secondary tabs when hovered may not meet WCAG AA requirements. (**EDIT:** Fixed, now they should.)

3. I’m unsure whether tabs should use the same styling as `.pf-m-current` when focused/clicked, or if they should continue to look like their hover state until JS finishes loading the tab content and adds `pf-m-current`. If JS ever fails to load the content the tab shouldn't look active, right? (**EDIT:** The custom focus state has been removed in favor of the eventual universal focus state.)

## Future notes

1. Scroll buttons should probably fade in and out when activated, but because `display: none;` is needed to guarantee that they don’t appear to screen readers, adding that animation is a bit tricky.

2. Page 2 of the design spec shows the tab separator stretching past the width of the content area while the tabs themselves stay within it. I’m not quite sure how this would be implemented. (**EDIT:** Probably some negative margin trickery. It'll be weird.)

patternfly-build commented:
Deploy preview for *pf-next* ready!

Built with commit a8a96f54833d2b938ac4a55f733486ba8672eef9

https://deploy-preview-868--pf-next.netlify.com

mcarrano commented:
This is looking good @andybraren But there are a few comments I have. First, to address your questions, yes, I agree that the hover state associated with secondary tabs is too light. @mceledonia can you take a look? Good question about whether we should have a "down" state between hover and selection. Do we have this as a distinct visual state for similar components (buttons for example)? I'll defer to Visual Design here.

In comparing the implementation to the visual design, I feel like the border around the tabs is a little heavy. @mceledonia what do you think? Maybe it's just the way the PDF renders but want to make sure the border weight and color is correct.

mceledonia commented:
A few hopefully quick visual changes, good eye @mcarrano .

**Border color**
#EDEDED 

**Spacers**
Above and below the text should be 8px spacers, just double checking that here.

**Primary tab hover state**
3px tall top-aligned rectangle, #8B8D8F, other borders still stay #EDEDED.

**Secondary tab hover state**
I agree this is too light. Let's use our secondary text color, I believe it might be $pf-global--Color--dark-200. It will be a lot more subtle but in use should be a clear change.

**Arrow button hover state**
Let's reuse the same hover state as primary tab hover state, having that same 3px rectangle appear at the top.

**Pressed/Down State**
I don't think we have this for other components. It may be something we want to explore, typically you see this as a shade darker than the hover state, or an inner-shadow appearing, etc... Material design has a very cool animated state for this. I don't think this state is crucial but could be nice to have if we want to explore it, though since we don't have it in other components I'm not sure of the best approach for adding it.

andybraren commented:
Thank you @mcarrano @mceledonia!

- Border color is now #EDEDED (is there a global variable for that color?)
- Spacers above and below tab text are 8px (`$pf-global--spacer--sm`)
- Primary tab hover state rectangle is now #8B8D8F
- Secondary tabs are now `$pf-global--Color--dark-200` on hover/focus
- Arrows now use the same hover state

The current pressed/down state is the same as the hover state. I can remove this if we want.

I have no idea if this is a possible scenario, but if there's ever a situation where the content of a tab is dependent on JavaScript/React pulling in stuff from elsewhere, some kind of state that indicates that the tab has been clicked but hasn't been populated yet might make sense. My thinking is that if the tab were to immediately switch to the regular unhovered/unclicked state, the user may interpret that as the tab "not working" the way that it should when it's actually just JS being a bit slow. Maybe it's a non-issue.

mcarrano commented:
Good point @andybraren . We will probably need some type of general "loading" state that can be applied in places like this. @LHinson do we have this in the backlog? If not I can open a new issue for this.

LHinson commented:
@mcarrano we have an issue for skeleton loading #745 and loading guidance #758 
I'm not sure if we should address this specifically for tab, rather just as a general rule for loading?

mcarrano commented:
I think this is pretty related to #758 @LHinson . Whatever we do there should carry into tabs. A new issue is not needed.

jgiardino commented:
The tabs look great!

In your original comment, you included this note:

> Scroll buttons should probably fade in and out when activated, but because display: none; is needed to guarantee that they don’t appear to screen readers, adding that animation is a bit tricky.

When testing with JAWS, I have noticed that `display:none` alone is not an effective way to hide elements from the screen reader. The most reliable way to [hide something from screen readers](https://a11yproject.com/posts/how-to-hide-content/) is with the attributes `hidden` or `aria-hidden=“true”`. Using just `display: none;` could result in some odd behavior. 

I’m copying the following from the Accessibility chapter in Smashing Book 6, from the section titled “Disabling Background Content”:

> Is the inactive content hidden from everyone?
>
> “Use CSS display: none. This turns the display off and disables all focusable items inside. If you have to animate it, you can transform opacity and set the display on the last keyframe of a CSS animation or after a setTimeout in JavaScript.”
>
> Excerpt From: Smashing Magazine. “Smashing Book 6: New Frontiers In Web Design.” iBooks. 

I found an example similar to what the author describes in [this medium article](https://medium.engineering/five-goofy-things-medium-did-that-break-accessibility-3bc804ae818d). 


P.S. After skimming through the article I mentioned, they suggest that you could effectively hide elements from screen readers using `display:none` and `visibility: hidden`. I can test this combination in JAWS later this week when I’m at home.


**Here are some other comments based on review:**
1. It isn’t obvious when a selected tab has keyboard focus.
2. The css removes default browser focus styles for tabs, but not for the scroll buttons.
2. The scroll buttons and tabs should be placed in the DOM in the same order that they appear on screen—just in case the screen reader user happens to come across them. Here were comments that I had posted in the original sequence PR:
    - I tested this [codepen](https://codepen.io/benfrain/full/zZZLaP/) in JAWS, which is similar to the design we're using. I noticed that the tabs will scroll to the left as JAWS focus moves through the tabs. Unfortunately, the tabs will not scroll to the right as JAWS focus moves through the tabs to the left. And the arrow buttons are visible to the screen reader, but not labeled and not in a DOM order that matches layout. They’re positioned at the end of the tabs, so leaving the last tab, I hear “unlabeled one button” and “unlabeled two button”. They’re not labeled for the screen reader or hidden from the screen reader.
    - My suggestion would be to place the buttons in the DOM to match the order they appear in the layout (e.g. start button before the list items and end button after the list items). This way, if the screen reader user does need them, they're present in a logical order.
3. This is probably a topic for our next a11y discussion, but... My suggestion would be to remove the role attributes from this component. Using the roles associated with tabs would require us to [manage focus](https://www.w3.org/TR/wai-aria-1.1/#managingfocus) as described by the authoring practices ([managing focus](https://www.w3.org/TR/wai-aria-practices-1.1/#kbd_roving_tabindex), [tabs examples](https://www.w3.org/TR/wai-aria-practices-1.1/#tabpanel)). 
    - I'm fine with holding off on changes related to this until we've had a chance to discuss as a team just to confirm what approach we want to take. In a recent meeting, we had discussed our approach to keyboard interaction. My recommendation is that for this initial release, we only focus on enabling the Esc key, and otherwise we rely on keyboard interaction that’s enabled by default (i.e. Tab key to shift focus) instead of managing focus as described in the resources I referenced above. My thinking here is that managing focus in a consistent way across all of our components might be more than we can handle for our initial release, and that this is an enhancement that we should be able to add later to our JS components without resulting in a breaking change.

LHinson commented:
@mattnolting will you help review this one as well?

andybraren commented:
Thank you for your comments @jgiardino !

0. Let me know if using `display:none;` and `visibility:hidden;` works just as well as adding a `hidden` attribute to the buttons in HTML. I imagine controlling that “automatically” through the simple addition of `.pf-m-start` or `.pf-m-end` would be preferred by devs, rather than remembering to add the `hidden` attribute.
     - I tried to do that keyframe trick but unfortunately couldn’t get the animation to be correct on both fade in and fade out. I’ll try again at a later date.
1. I agree. @mceledonia  is our intent still to override the default browser keyboard focus styling with our own? If so, active tabs (like “Tab item 2” at the [very top of the preview](https://deploy-preview-868--pf-next.netlify.com/components/Tabs/examples/)) need a focus state that doesn’t rely on the top rectangle’s color (since active tabs already have one).
     - If this is a problem we’d prefer to revisit later on, I can re-enable the browser’s default focus style for now.
2. I’ve removed the focus style of scroll buttons for consistency’s sake, but active scroll buttons have the same problem as 1.
3. The DOM now matches the visual order. Sub-bullet 1 is referring to that Pen and not this implementation, right? Our buttons have `aria-label`, and I don’t have JAWS so I can’t test the scrolling behavior you describe.
4. I’ve removed `role=“tab”`, `role=“tablist”`, `role=“tabpanel”`, and `role=“presentation”`. I based those attributes on [Inclusive Components’ example](https://inclusive-components.design/tabbed-interfaces/).

jgiardino commented:
Sounds good, Andy!

> Sub-bullet 1 is referring to that Pen and not this implementation, right?

Correct. I borrowed some comments I made when testing the Sequence component. 😉 I can do a full screen reader review tomorrow when I'm working from home. I'd also like to test some other design system implementations of tabs, too.

That's awesome that you read the Inclusive Components article on tabs! I was thinking about making that required reading for the next a11y discussion.  He makes some good points that I'd love to get the teams thoughts on.

mceledonia commented:
This looks like a good time to try our custom focus state. Let's see how it works! You can find a visual reference in #790 

Spec:
8px padding on each side of the tab 
Fill is #007BBA at 15% opacity 
Border is #007BBA at 100% opacity, 1px width, 3px radius

andybraren commented:
Note: I've removed this component's custom focus states entirely and it now uses the browser's default outline instead. Once #790 is implemented our new styling should apply to these tabs as well.

andybraren commented:
@jgiardino Just fyi, I've updated this PR and its documentation to include the `hidden` attribute on scroll buttons whenever they should be ignored by screen readers. I'd still be interested in the viability of a CSS-only solution though (`display:none;` and `visibility:hidden;`) to make the React devs' lives easier. 🙂

matthewcarleton commented on src/patternfly/components/Tabs/tabs.scss:
```diff
@@ -0,0 +1,189 @@
+  background-color: var(--pf-c-tabs__scroll-button--BackgroundColor);
+  border: var(--pf-c-tabs__item--BorderWidth) solid var(--pf-c-tabs__scroll-button--BorderColor);
+  border-radius: initial;
+
```
if you set the color here the theming should work.

matthewcarleton replied:
right so because you aren't specifying a colour for the icon it's just taking the default and when the theme is toggled anything dark goes to light. So you'd have to set the colour of the explicitly here to override that for the theme. Does that help?

andybraren replied:
Confusion is gone and `--pf-global--BackgroundColor--100` now adjusts to dark themes. Thanks for figuring that out! 👍 

Cards, Nav, and Popover are the only other components using `--pf-global--BackgroundColor--100` that could be affected when the fix included in this PR is merged.

- Card has custom styling within `placeholders.scss` that switches its background to `--pf-global--BackgroundColor--dark-transparent-200` while in dark mode. It probably shouldn't be doing that in there, but that's its own problem.
- Nav sets a background color variable, but then never actually uses it, so that's fine.
- Popover had the same issue I did with this Tabs component and it now switches to a dark background, but it's barely visible and should probably switch to using the same `--pf-global--BackgroundColor--dark-transparent-200` that Card does to look okay.

matthewcarleton commented on src/patternfly/components/Tabs/tabs.scss:
```diff
@@ -0,0 +1,189 @@
+
+  // Primary Tab Items
+  --pf-c-tabs__item--BackgroundColor:       var(--pf-global--BackgroundColor--100);
+  --pf-c-tabs__item--current--Color:        var(--pf-global--active-color--100);
```
switching this to `--pf-global--primary-color--100` fixes the theme issue but I think we need to investigate this further.

andybraren commented on src/patternfly/components/Tabs/docs/code.md:
```diff
@@ -0,0 +1,7 @@
+## Overview
+
+`Tabs` should only be used to change content views within a page. The similar-looking but semantically different  `Nav` component is available for navigation use cases.
```
Done and done.

andybraren commented on src/patternfly/components/Tabs/examples/tabs-secondary.hbs:
```diff
@@ -0,0 +1,13 @@
+<div class="pf-c-tabs-container">
+  {{#> tabs tabs--attribute='aria-label="Secondary Tabs Example"'}}
+    {{#> tabs-item tabs-item--attribute='id="ex3-tab1"'}}Tab item 1{{/tabs-item}}
+    {{#> tabs-item tabs-item--attribute='id="ex3-tab2"' tabs-item--current="true"}}Tab item 2{{/tabs-item}}
```
Do you mean making "Secondary tab item 1" have `.pf-m-current`? That makes sense (done).

Do you think we should also make every "Tab item 1" have `.pf-m-current` rather than "Tab item 2" across all three examples? I suppose that would also be the typical default.

matthewcarleton replied:
👍

andybraren replied:
👍

andybraren commented on src/patternfly/components/Tabs/tabs.scss:
```diff
@@ -0,0 +1,189 @@
+  // Primary Tab Items
+  --pf-c-tabs__item--BackgroundColor:       var(--pf-global--BackgroundColor--100);
+  --pf-c-tabs__item--current--Color:        var(--pf-global--active-color--100);
+  --pf-c-tabs__item--hover--BoxShadowColor: #8b8d8f;
```
I've added a commit with the revised box-shadow variable implementation. Let me know if the variable names are correct (`--current` isn't really a state, but maybe it kinda is?).

@mceledonia specified these colors [above](https://github.com/patternfly/patternfly-next/pull/868#issuecomment-434448363), and to my knowledge they aren't global variables. Not sure how to proceed with that.

I'll work on that visual issue. The 45 degree angle seems to be interpolation resulting from a single same-colored pixel in that corner set by `border-top-color`. I'll need to tweak my box-shadow mimicry. Good eye! 🙂

matthewcarleton replied:
awesome! Can you confirm with @mceledonia about that. I don't understand why we would want to introduce a new border colour just for tabs.

andybraren replied:
Border colors resolved, now just need to fix the glitch (hopefully by switching to ::before)

andybraren replied:
Drop shadows have been dropped entirely and tabs now use borders on `::before` and `::after` to achieve the same result while being visible within Windows' high contrast mode. This required a lot of SCSS restructuring and eliminated the need for some variables entirely. Let me know if the new SCSS looks good and makes sense.

andybraren replied:
Seems like we're all satisfied with this.

andybraren commented on src/patternfly/components/Tabs/examples/tabs-primary.hbs:
```diff
@@ -0,0 +1,15 @@
+  {{#> tabs-item tabs-item--attribute='id="ex1-tab3"'}}Tab item 3{{/tabs-item}}
+{{/tabs}}
+
+<section id="section1" aria-labelledby="ex1-tab1" hidden>
```
I've added `tabs-section.hbs` and it's now used to generate the sections within the Primary tabs example. Do you think sections should be included in every example?

andybraren commented on src/patternfly/components/Tabs/tabs.scss:
```diff
@@ -0,0 +1,189 @@
+
+// Tabs
+.pf-c-tabs {
+  /* stylelint-disable */
```
Done, and for good reasons, although some components still line up property values for legibility's sake (see `table.scss`)?

mcoker replied:
We should decide if we want to maintain doing that or not, so all of the scss files are the same. Want to bring it up in the css army or our next meeting?

jgiardino commented:
Hey Andy, here are some notes based on review with screen readers.

The only suggestion I have at this time is to add `aria-controls` to the tabs. This is noted in the [wai-aria authoring practices for Tabs](https://www.w3.org/TR/wai-aria-practices-1.1/examples/tabs/tabs-1/tabs.html). I included a comment about this below, too.

**Using `display:none; visibility:hidden;` to hide elements from screen readers**
This worked in the three screen reader / browser combinations that we test in.

**Testing other implementations**

1. [Carbon design system ](https://www.carbondesignsystem.com/components/tabs/code)
    - Implementation: use roles, but do not manage focus as prescribed by wai-aria. They do implement left/right arrow keys, but the Tab key also works
    - Results:
      - Works in JAWS and VO
      - It doesn't work as expected in NVDA, which may or may not be the result of using roles without implementing the keyboard behavior. When I shift focus to the tabs, it announces all tabs at once. The only way to shift focus is by using the Tab key.
      - NOTE: They also use `aria-controls` which only seems to be supported in JAWS, but since it is recommended in the wai-aria guide, is probably good to include.
2. [Ant design system](https://ant.design/components/tabs/)
    - Here I was mainly testing their scrolling example, to see how it works with screen readers
    - Results:
      - Fails in both JAWS and NVDA, in so many ways...
3. [Material design system](https://material-components.github.io/material-components-web-catalog/#/component/tabs)
    - I was also mainly testing their scrolling example, to see how it works with screen readers. 
    - Results:
      - Works reasonably well in both JAWS and NVDA. There were times in JAWS where a tab with focus wasn't visible (i.e. the edge was visible but not the label, and therefore it didn't scroll over)
      - This implementation does not include scroll buttons, which I could see as being problematic for some users.

mcoker commented on src/patternfly/components/Tabs/docs/tabs-primary.md:
```diff
@@ -0,0 +1,12 @@
+
+| Attribute | Applied To | Outcome |
+| -- | -- | -- |
+| `id="tab1"` | `.pf-c-tabs__item` | Uniquely identifies the tab. **Required** |
```
should this page have the base classes to create a tabs component?

andybraren replied:
Added `.pf-c-tabs`, `.pf-c-tabs__scroll-button`, `.pf-c-tabs__list`, `.pf-c-tabs__item`, and `.pf-c-tabs__button` to the table. Let me know if that looks correct.

mcoker commented on src/patternfly/components/Tabs/tabs-section.hbs:
```diff
@@ -0,0 +1,6 @@
+<section class="pf-c-tabs__section"{{#if badge--modifier}} {{badge--modifier}}{{/if}}
```
```suggestion
<section class="pf-c-tabs__section"{{#if tabs-section--modifier}} {{tabs-section--modifier}}{{/if}}
```

andybraren replied:
Oops, thanks for spotting that, fixed.

mcoker commented on src/patternfly/components/Tabs/tabs.hbs:
```diff
@@ -0,0 +1,20 @@
+<div class="pf-c-tabs{{#if tabs--modifier}} {{tabs--modifier}}{{/if}}">
```
can you add a block for `{{tabs--attribute}}`?

andybraren replied:
Added.

mcoker commented on src/patternfly/components/Tabs/tabs.scss:
```diff
@@ -0,0 +1,188 @@
+.pf-c-tabs {
+  // Primary Tab Items
+  --pf-c-tabs__item--BackgroundColor: var(--pf-global--BackgroundColor--100);
+  --pf-c-tabs__item--current--Color: var(--pf-global--active-color--100);
```
```suggestion
  --pf-c-tabs__item--m-current--Color: var(--pf-global--active-color--100);
```

mcoker replied:
just update the rest of the scss to use `m-current` in the var name since it's a modifier class

andybraren replied:
Added the `m-` to `--pf-c-tabs__item--m-current--Color`, `--pf-c-tabs__item--m-current--BoxShadow`, and `--pf-c-tabs__item--m-current--ZIndex`. Thanks!

mcoker commented on src/patternfly/components/Tabs/tabs.scss:
```diff
@@ -0,0 +1,188 @@
+  --pf-c-tabs__item--hover--BoxShadow: inset 0 pf-size-prem(2) 0 0 var(--pf-c-tabs__item--hover--BoxShadowColor);
+  --pf-c-tabs__item--current--BoxShadow: inset 0 pf-size-prem(2) 0 0 var(--pf-c-tabs__item--current--Color);
+  --pf-c-tabs__item--BorderWidth: var(--pf-global--BorderWidth--sm);
+  --pf-c-tabs__item--BorderColor: #ededed;
```
should this be

```suggestion
  --pf-c-tabs__item--BorderColor: var(--pf-global--BorderColor--light-200);
```

andybraren replied:
It should, and now I've discovered `colors.scss`. Thanks!

Would doing this for the box shadow color be legit? Right now it's set to `#8b8d8f` which isn't within `scss-variables.scss`.

    --pf-c-tabs__item--hover--BoxShadowColor: $pf-color-black-500;

mcoker replied:
Ideally we should have a CSS var for it if we're going to use it in the component. Can you use one of `--pf-global--BorderColor--dark` or `--pf-global--BorderColor--light` instead? I wonder what similar components with this color on a hover state?

andybraren replied:
I'm gonna go with `--pf-global--BorderColor--light` which translates to `$pf-color-black-400` if @mceledonia is okay with it - just `100` shy of the intended `#ededed` (`$pf-color-black-500`).

`--dark` would make it `$pf-color-black-600` which is probably fine too.

andybraren replied:
Got approval from @mceledonia, good to go with `--pf-global--BorderColor--light`. 👍

mcoker commented on src/patternfly/components/Tabs/tabs.scss:
```diff
@@ -0,0 +1,188 @@
+  // Filled style
+  &.pf-m-fill {
+    display: flex;
+    width: 100%;
```
I think you can just use `display: flex;` or `width: 100%;` here? I would probably use `display: flex;`

andybraren replied:
You're right, I've removed the `width: 100%;`.

mcoker commented on src/patternfly/components/Tabs/tabs.scss:
```diff
@@ -0,0 +1,189 @@
+// Tab Items
+.pf-c-tabs__item {
+  flex: none;
+  border-top: var(--pf-c-tabs__item--BorderWidth) solid var(--pf-c-tabs__item--BorderColor);
```
I might also rewrite this as `border: solid var(--pf-c-tabs__item--BorderColor); border-width: var(--pf-c-tabs__item--BorderTopWidth) var(--pf-c-tabs__item--BorderRightWidth)...`

andybraren replied:
Done and done (hopefully correctly). Thank you both.

mcoker commented on src/patternfly/components/Tabs/tabs.scss:
```diff
@@ -0,0 +1,188 @@
+  // Different styling for secondary tab buttons
+  .pf-c-tabs__button {
+    box-shadow: initial;
+    box-shadow: none;
```
do you need both of these?

andybraren replied:
Nope. 👍

mcoker commented on src/patternfly/sass-utilities/placeholders.scss:
```diff
@@ -2,10 +2,9 @@
-  --pf-global--active-color--100: var(--pf-global--active-color--100);
   --pf-global--primary-color--100: var(--pf-global--primary-color--dark-100);
   --pf-global--link--Color: var(--pf-global--primary-color--dark-100);
+  --pf-global--BackgroundColor--100: var(--pf-global--BackgroundColor--light-100);
```
🔥

andybraren replied:
I got permission first 👍

andybraren commented on src/patternfly/components/Tabs/tabs.scss:
```diff
@@ -0,0 +1,189 @@
+  }
+
+  // Scroll button modifiers
+  &.pf-m-start .pf-c-tabs__scroll-button:nth-of-type(1) {
```
I've condensed them. I think `nth-of-type` makes sense, since there should only be two scroll buttons of the class `.pf-c-tabs__scroll-button` within the direct parent container, right?

andybraren commented on src/patternfly/components/Tabs/tabs.scss:
```diff
@@ -0,0 +1,189 @@
+  position: absolute;
+  top: 0;
+  bottom: 0;
+  z-index: 2;
```
Done.

andybraren commented:
Note to self: @mceledonia noticed the box-shadow appearing within secondary tabs when hovering over that region, and I noticed them flickering in and out when wiggling over that area in Chrome. That should be figured out.

andybraren commented:
@jgiardino

Thank you Jenn! Sounds like we can avoid requiring devs to add `hidden` to `.pf-c-tabs__scroll-button` then and achieve the same screen reader experience with CSS using `.pf-m-start` and `.pf-m-end`. Nice!

Regarding this:

> The only suggestion I have at this time is to add `aria-controls` to the tabs. This is noted in the [wai-aria authoring practices for Tabs](https://www.w3.org/TR/wai-aria-practices-1.1/examples/tabs/tabs-1/tabs.html). I included a comment about this below, too.

That links mentions that `aria-controls` "Refers to the `tabpanel` element associated with the tab." I removed `role=“tabpanel”` from the `<section>` elements in a previous commit. Should I add those back in while I'm adding `aria-controls` to each tab?

jgiardino commented:
I'm still hesitant to add the tab roles back without committing to implementing the keyboard behavior as documented by wai-aria. 

[aria-controls](https://www.w3.org/TR/wai-aria-1.1/#aria-controls) can be used for any role. I only mention the [wai-aria AP for Tabs](https://www.w3.org/TR/wai-aria-practices-1.1/#tabpanel) because I was being lazy and was trying to borrow their documentation on how to apply it, since we have a similar structure.

It's worth mentioning that `aria-controls` is only supported by JAWS, and Heydon Pickering wrote an [article](http://www.heydonworks.com/article/aria-controls-is-poop) explaining why we shouldn't use it. But in the comments, Léonie Watson makes some valid points in support of using it.

It's also worth mentioning that in [Heydon's article on Tabs](https://inclusive-components.design/tabbed-interfaces/), he suggests that a non-aria version of Tabs is to create the "table of contents" pattern that uses anchor links to access the sections of content. But since we ultimately plan to add the roles and keyboard interaction in, I feel like what you have now gets us closer to our ultimate goal.

andybraren commented:
@jgiardino Thank you for clarifying, and I agree `aria-controls` is worth adding for the reasons Léonie mentions. I'll add that in soon. Thanks again! 👍

andybraren commented on src/patternfly/components/Tabs/examples/tabs-secondary.hbs:
```diff
@@ -0,0 +1,13 @@
+<div class="pf-c-tabs-container">
```
I agree, `pf-c-tabs` is now the container and buttons are now within lists. This required a lot of restructuring of the handlebars files, so let me know if it looks okay.

andybraren replied:
FYI, the current implementation has some positioning issues. I'm not quite sure the buttons can be stickied/fixed to the left and right while being within `.pf-c-tabs__list`, but I'll keep trying.

andybraren replied:
Thanks to @mcoker and @mattnolting I think we have a solution.

Tabs never need a container, whether or not secondary tabs are present. Instead, in the specific case where secondary tabs are directly adjacent to primary tabs, we use an adjacent sibling selector to style those secondary tabs appropriately (adding top border, subtracting margin to prevent stacked borders, setting `display:flex;` to move secondary tabs to a new line).

This seems to meet the design requirements. One side effect of setting secondary tabs to be full width without a container is that the right-most scroll button is positioned way off to the right. Since scroll buttons are conditionally activated with JavaScript, I'm not too concerned about that.

<img width="812" alt="secondary-tabs-right-scroll-button" src="https://user-images.githubusercontent.com/9122899/48490591-c13efd00-e7f3-11e8-8a28-d44f07eaf015.png">

jgiardino replied:
I'm not sure we can rely on secondary tabs being adjacent siblings of primary tabs.

When I select a primary tab, the tab panel associated with that selected primary tab becomes visible. Wouldn't the secondary tabs live in that tab panel?

andybraren replied:
@jgiardino We only rely on secondary tabs (or the tab panel that contains them) being directly adjacent to the primary tabs if the styling in the example is what the dev is looking for (with primary tabs on top and secondary directly below them, which is probably pretty common, and the DOM should reflect that positioning).

Secondary tabs that are revealed by primary tabs could be somewhere completely separate in the UI, which isn't reflected in the design doc but is something that some products do (@mattnolting can attest). In that case our Primary + Adjacent Secondary styling won't apply.

Should the markup of our secondary tabs example be updated to have a container that has `role="tabpanel"`, or maybe putting that `role` directly onto `.pf-m-tabs-secondary` would work?

andybraren replied:
I tweaked this so that the secondary tab component can be buried within a div and still receive the right styling, so long as that containing div has `.pf-m-tabs-secondary` and is directly adjacent to the primary tabs. React devs feel confident that they won't need containing divs, but it's there if needed.

The `.pf-m-tabs-secondary` modifier is still a bit weird because it changes the component's margins, but only in a highly-specific context that's "conceptually within the rules" even though it technically breaks them. The component itself doesn't have margins and can be dropped anywhere and work as usual, which we seem to be okay with.

andybraren commented:
@mcarrano "Filled tabs" has been fixed, thank you for spotting that.

jgiardino commented:
I was just taking a quick look again at your latest updates, and noticed that you have a `<button>` element as a child of `<ul>`, which I think is invalid?

Also, a question that would be good to explore in the future is the relationship of the secondary tabs to the primary tabs. Would/should the secondary tabs be included within the tab panel of the primary tab? I think it's totally fine to leave the secondary tabs example as-is for now, and then later revisit it if we find that the react example is using a different html structure.

Great job on this one!!

andybraren commented:
@jgiardino Agreed, this is one of the last things to fix/figure out. I think @matthewcarleton wants to discuss component structuring with the larger group (should the first element always be the top-level component class?) and we could probably discuss secondary tab location as well.

Thanks, and thank you and everyone else who reviewed for helping me fix things. 👍

mattnolting commented on src/patternfly/components/Tabs/examples/tabs-secondary.hbs:
```diff
@@ -0,0 +1,39 @@
+        Tab item 3
+      {{/tabs-button}}
+    {{/tabs-item}}
+  {{/tabs-list}}
```
Might be easier to understand if you make the tab titles longer and/or add several more. It would help illustrate the scroll buttons appearing only when tabs reach the overflow point.

andybraren replied:
This particular code chunk is for the secondary tabs example though, with scroll buttons off. I have 6 tab items in the scroll buttons example - do you think I should add more there?

mattnolting replied:
Yes, sorry. I put the comment on the wrong file.

andybraren replied:
Done.

andybraren commented on src/patternfly/components/Tabs/tabs.scss:
```diff
@@ -0,0 +1,273 @@
+  --pf-c-tabs__scroll-button--ZIndex: var(--pf-global--ZIndex--md);
+
+  position: relative;
+  display: inline-flex;
```
Hmm. I guess it's a bit hypocritical to have primary and secondary be different, but the scroll buttons example (with scroll buttons manually turned on without JS) will look kinda broken. We're good with that?

mattnolting replied:
If you add more text to tabs (and possibly more tabs) to the scroll example, you can show the scroll buttons in the proper context.

![screen shot 2018-11-14 at 11 30 30 am](https://user-images.githubusercontent.com/5385435/48496753-bb9be400-e800-11e8-9235-6c1ffa67b538.png)

andybraren replied:
The scrolling example now has 10 tabs (same text though). The workspace now has an issue where some tabs begin to overflow the box if the mobile view is selected.

andybraren commented on src/patternfly/components/Tabs/tabs.scss:
```diff
@@ -0,0 +1,273 @@
+
+  // Primary with secondary tabs style
+  & + .pf-m-tabs-secondary {
+    display: flex;
```
Done.
